Added endpoint to translate sections to wikitext (T94890).

Pchelolo opened the pull request:
A new endpoint /transform/sections/to/wikitext/{title}/{revision}
was added. It takes the original revision from restbase, replaces
sections according to JSON, provided in a sections parameter, and
posts it to parsoid for html->wikitext translation.

A set of tests was added verifying the sections could be replaced
and added, and verifying errors on non-existent sections and invalid
JSON.

Bug: https://phabricator.wikimedia.org/T94890

Pchelolo commented on mods/parsoid.js:
```diff
@@ -530,6 +551,45 @@ function cheapBodyInnerHTML(html) {
+    var sectionOffsets = original['data-parsoid'].body.sectionOffsets,
+        newBody = cheapBodyInnerHTML(original.html.body);
+    Object.keys(sectionsJson)
+    .sort(function(id1, id2) {
```
We need to sort section in reverse order by section offset so that replaces don't clash with offset numbers.

d00rman replied:
That's correct :+1: But, please, let's not confuse the reader / maintainer of the code - use `offset1` for `id1` (idem for `offset2`) in the line below.

d00rman replied:
Also, `.sort()` can (and should) be moved to the line above

Pchelolo commented on mods/parsoid.js:
```diff
@@ -530,6 +551,45 @@ function cheapBodyInnerHTML(html) {
+            + sectionsJson[id]
+            + newBody.substring(offset.html[1], newBody.length);
+    });
+    return '<body>' + newBody + '</body>';
```
According to @gwicke parsoid html->wikitext translation returns identical results for full html and only body when body is wrapped with <body>. The 'newBody' here was created by cheapBodyInnerHTML, so it for sure not wrapped yet.

d00rman commented on mods/parsoid.js:
```diff
@@ -530,6 +551,45 @@ function cheapBodyInnerHTML(html) {
+ */
+function replaceSections(original, sectionsJson) {
+    var sectionOffsets = original['data-parsoid'].body.sectionOffsets,
+        newBody = cheapBodyInnerHTML(original.html.body);
```
No variable def chaining please, use `var newBody = ...`

d00rman commented on mods/parsoid.js:
```diff
@@ -530,6 +551,45 @@ function cheapBodyInnerHTML(html) {
+    .forEach(function(id) {
+        var offset = sectionOffsets[id];
+        if (!offset) {
+            throw new rbUtil.HTTPError({
```
This has already been checked during the sort above, hasn't it?

Pchelolo replied:
@d00rman In case there's only 1 item in the sort comparator function wouldn't be called.

d00rman replied:
> Alternatively we can sort without check and put undefined to the beginning, and then check only in one place.

Yup, that. Or use `Array.some()` to find out if there are any undefs before iterating so have a clearer flow/check separation.

d00rman commented on specs/mediawiki/v1/content.yaml:
```diff
@@ -760,6 +760,52 @@ paths:
+          description: Sections to transform
+          type: string
+          required: true
+      responses:
```
Missing the `400` response definition

d00rman commented on mods/parsoid.js:
```diff
@@ -530,6 +551,45 @@ function cheapBodyInnerHTML(html) {
+    Object.keys(sectionsJson)
+    .sort(function(id1, id2) {
+        var offset1 = sectionOffsets[id2],
+            offset2 = sectionOffsets[id1];
```
Also variable def separation :)

d00rman commented on test/features/parsoid/transform.js:
```diff
@@ -143,8 +143,47 @@ describe('transform api', function() {
+        .then(function(res) {
+            assert.deepEqual(res.status, 200);
+            assert.contentType(res, 'text/plain;profile=mediawiki.org/specs/wikitext/1.0.0');
+            assert.deepEqual(/== First section replaced ==/.test(res.body), true);
```
Should also probably check that `== First section ==` (or whichever title was there before) is not part of the body any more.

d00rman commented:
LGTM :+1: A couple of small nits in-lined.

Pchelolo commented:
Already fixing those.

Pchelolo commented:
@d00rman Fixed. Also I've changed test to target a special subpage I've created and marked as 'DO NOT MODIFY'

d00rman commented on mods/parsoid.js:
```diff
@@ -459,7 +460,27 @@ PSP.transformRevision = function (restbase, req, from, to) {
-        body2[from] = req.body[from];
+        if (from === 'sections') {
+            try {
+                var sections = JSON.parse(req.body.sections);
```
Two small remarks:

- You might want to check that `req.body.sections` is not already something that's not a String or Buffer before parsing it, which it will be in case the client supplies the `application/json` *Content-Type* header.
- Probably only the JSON parsing should be enclosed in a `try/catch` block, since `replaceSections` is going to throw its own errors.

d00rman replied:
> You might want to check that `req.body.sections` is not already something that's not a String or Buffer

This also implies that it's safer to do:

```js
var sections = JSON.parse(req.body.sections.toString())
```

when we indeed need to parse it

Pchelolo replied:
@d00rman Actually, this endpoint is made in analogy to other transformation endpoints, so they accept a form-data. So, in case application/json header is defined it will fail way before this point with 400 'Missing request parameter sections'. So, I suppose it's enough to add .toString() and check that we've got an object after parsing to protect against smth like sections=1234 where JSON.parse wouldn't throw an exception. What do you think?

d00rman replied:
Sorry, I probably wasn't clear (and now re-reading my own comment proves it). I just meant that it'd be nice to be flexible so that in case the content type is changed to `application/json` for the code to still be working (I am basing this possibility on the fact that it only takes to change one line in the spec).

> So, I suppose it's enough to add .toString() and check that we've got an object after parsing to protect against smth like sections=1234 where JSON.parse wouldn't throw an exception.

I must admit I don't understand what you mean here. How would `"sections=1234".toString()` change anything?

Pchelolo replied:
Ah, ok, I'll add JSON handling. 

Regarding your second comment, it would be not `'sections=1234'.toString()` but `1234.toString()` but anyway this is not relevant now. I'll reupload the PR after I work out son handling.

Pchelolo commented:
@d00rman I've added a check that body.sections is not already an object before parsing it. In case of JSNO content-type parsing would be skipped, in case of form-data we would parse the string representation anyway, in case wrong JSON is provided within sections we would get 400. A test is also added
